README.md


[![Build Status](https://travis-ci.org/fmichonneau/phylobase.png?branch=master)](https://travis-ci.org/fmichonneau/phylobase.png)

Latest Build log: https://travis-ci.org/fmichonneau/phylobase

# phylobase

## About this package

`phylobase` provides classes and methods to easily associate, manipulate,
explore, and plot phylogenetic trees and data about the species they
include. The goal of this package is to provide a base set of tools likely to be
shared by all packages designed for phylogenetic analysis. This standardization
will benefit both *end-users* by allowing them to move results across packages
and keep data associated with the phylogenetic trees; and *developers* by
focusing on method development instead of having to rewrite the base functions.

- Authors: R Hackathon et al. (alphabetically: Ben Bolker, Marguerite Butler,
  Peter Cowan, Damien de Vienne, Dirk Eddelbuettel, Mark Holder, Thibaut
  Jombart, Steve Kembel, Francois Michonneau, David Orme, Brian O'Meara,
  Emmanuel Paradis, Jim Regetz, Derrick Zwickl)
- Maintainer: Francois Michonneau
- Licence: GPL (>= 2)
- Issues, bug reports, feature requests, discussion:
  http://github.com/fmichonneau/phylobase/issues

## Installation

### Stable version

The stable version (the minor and patch version numbers are even, e.g., 0.6.8)
can be downloaded from CRAN.


```r
install.packages("phylobase")
library(phylobase)
```

### Development version

The development version (the patch version number is odd, e.g., 0.6.9) is
available on GitHub (https://github.com/fmichonneau/phylobase), and can be
installed using the `[devtools](http://cran.r-project.org/package=devtools)`
package.


```r
library(devtools)
install_github("phylobase", "fmichonneau")
library(phylobase)
```

### Getting started



`phylobase` comes with example data sets `geospiza` and `geospiza_raw`.

- `geospiza` is a `phylo4d` object (the `phylobase` class that holds together a
  phylogenetic tree and associated data, the `phylo4` class is for phylogenetic
  trees only).
- `geospiza_raw` is a list that contains the tree `geospiza_raw$tree` (as an
  `ape::phylo` object) and the data `geospiza_raw$data` (as a `data.frame`) that
  were used to build the `geospiza` object.

Now we'll take the \emph{Geospiza} data from \verb+geospiza_raw$data+ and merge
it with the tree. However, since \emph{G. olivacea} is included in the tree but
not in the data set, we will initially run into some trouble:


```r
data(geospiza_raw)
g1 <- as(geospiza_raw$tree, "phylo4")
geodata <- geospiza_raw$data
g2 <- phylo4d(g1, geodata)
```

```
## Error: The following nodes are not found in the dataset: olivacea
```

To deal with _G. olivacea_ missing from the data, we have a few choices. The
easiest is to use `missing.data="warn"` to allow `R` to create the new object
with a warning (you can also use `missing.data="OK"` to proceed without
warnings):


```r
g2 <- phylo4d(g1, geodata, missing.data="warn")
```

```
## Warning: The following nodes are not found in the dataset: olivacea
```

```r
head(g2)
```

```
##           label node ancestor edge.length node.type wingL tarsusL culmenL
## 1    fuliginosa    1       24     0.05500       tip 4.133   2.807   2.095
## 2        fortis    2       24     0.05500       tip 4.244   2.895   2.407
## 3  magnirostris    3       23     0.11000       tip 4.404   3.039   2.725
## 4   conirostris    4       22     0.18333       tip 4.350   2.984   2.654
## 5      scandens    5       21     0.19250       tip 4.261   2.929   2.622
## 6    difficilis    6       20     0.22800       tip 4.224   2.899   2.277
## 7       pallida    7       25     0.08667       tip 4.265   3.089   2.430
## 8      parvulus    8       27     0.02000       tip 4.132   2.973   1.974
## 9    psittacula    9       27     0.02000       tip 4.235   3.049   2.260
## 10       pauper   10       26     0.03500       tip 4.232   3.036   2.187
## 11   Platyspiza   11       18     0.46550       tip 4.420   3.271   2.331
## 12        fusca   12       17     0.53409       tip 3.975   2.937   2.052
## 13 Pinaroloxias   13       16     0.58333       tip 4.189   2.980   2.311
## 14     olivacea   14       15     0.88077       tip    NA      NA      NA
## 15         <NA>   15        0          NA      root    NA      NA      NA
## 16         <NA>   16       15     0.29744  internal    NA      NA      NA
## 17         <NA>   17       16     0.04924  internal    NA      NA      NA
## 18         <NA>   18       17     0.06859  internal    NA      NA      NA
## 19         <NA>   19       18     0.13404  internal    NA      NA      NA
## 20         <NA>   20       19     0.10346  internal    NA      NA      NA
##    beakD gonysW
## 1  1.941  1.845
## 2  2.363  2.222
## 3  2.824  2.676
## 4  2.514  2.360
## 5  2.145  2.037
## 6  2.011  1.930
## 7  2.016  1.949
## 8  1.874  1.813
## 9  2.230  2.074
## 10 2.073  1.962
## 11 2.347  2.282
## 12 1.191  1.401
## 13 1.548  1.630
## 14    NA     NA
## 15    NA     NA
## 16    NA     NA
## 17    NA     NA
## 18    NA     NA
## 19    NA     NA
## 20    NA     NA
```

### Importing data

#### From NEXUS files

`phylobase` has a robust parser for NEXUS files (it uses the NEXUS Class Library
from Paul Lewis and Mark Holder,
[NCL](http://sourceforge.net/projects/ncl/files/)). It can be used to import
simultaneously tree and species data.


```r
myrmeFile <- system.file("nexusfiles/treeWithDiscreteData.nex", package="phylobase")
myrme <- readNexus(file=myrmeFile)
head(myrme)
```

```
##                      label node ancestor edge.length node.type        time
## 1   Myrmecocystussemirufus    1       27       1.725       tip     diurnal
## 2   Myrmecocystusplacodops    2       27       1.725       tip     diurnal
## 3      Myrmecocystusmendax    3       26       4.651       tip     diurnal
## 4    Myrmecocystuskathjuli    4       28       1.084       tip     diurnal
## 5    Myrmecocystuswheeleri    5       28       1.084       tip     diurnal
## 6     Myrmecocystusmimicus    6       30       2.709       tip     diurnal
## 7     Myrmecocystusdepilis    7       30       2.709       tip     diurnal
## 8    Myrmecocystusromainei    8       32       2.194       tip     diurnal
## 9  Myrmecocystusnequazcatl    9       32       2.194       tip     diurnal
## 10       Myrmecocystusyuma   10       31       4.451       tip crepuscular
## 11   Myrmecocystuskennedyi   11       23       6.045       tip     diurnal
## 12 Myrmecocystuscreightoni   12       22      10.569       tip crepuscular
## 13  Myrmecocystussnellingi   13       33       2.770       tip crepuscular
## 14 Myrmecocystustenuinodis   14       33       2.770       tip crepuscular
## 15  Myrmecocystustestaceus   15       20      12.301       tip crepuscular
## 16  Myrmecocystusmexicanus   16       34       5.725       tip   nocturnal
## 17   Myrmecocystuscfnavajo   17       35       2.870       tip   nocturnal
## 18     Myrmecocystusnavajo   18       35       2.870       tip   nocturnal
## 19                    <NA>   19        0          NA      root        <NA>
## 20                    <NA>   20       19       1.699  internal        <NA>
##         subgenus
## 1   Endiodioctes
## 2   Endiodioctes
## 3   Endiodioctes
## 4   Endiodioctes
## 5   Endiodioctes
## 6   Endiodioctes
## 7   Endiodioctes
## 8   Endiodioctes
## 9   Endiodioctes
## 10  Eremnocystus
## 11  Endiodioctes
## 12  Eremnocystus
## 13  Eremnocystus
## 14  Eremnocystus
## 15 Myrmecocystus
## 16 Myrmecocystus
## 17 Myrmecocystus
## 18 Myrmecocystus
## 19          <NA>
## 20          <NA>
```

#### From NeXML


```r
library(RNeXML)
nxmlFile <- system.file("nexmlfiles/comp_analysis.xml", package="phylobase")
nxml <- nexml_read(nxmlFile)
nxmlEx <- phylo4(nxml)
```


NEWS



        *************************************************
        *                                               *
        *               0.6 SERIES NEWS                 *
        *                                               *
        *************************************************

                CHANGES IN phylobase VERSION 0.6.10

   *  Fix bug NA is labels were considered duplicated by checkPhylo4().

   *  Fix bug #605 (R-forge) -- treePlot subsets numeric data for plotting.

   *  Initial basic support for converting RNeXML objects in phylo4 and phylo4d
      format.

   *  All documentation is now in Roxygen format

   *  All tests done with testthat

   *  New methods: internalEdges(), terminalEdges()

   *  hasPoly, hasRetic, hasSingle are now methods instead of functions.

   *  rootNode returns the rootNode using the same format as getNode().

                CHANGES IN phylobase VERSION 0.6.8

   *   Not many user-visible changes, most are related to improving speeds
       during test of object validation (most tests done in C++) and to getNode
       that is used by many functions.

   *  Changes to package structure to make it compatible with devtools
      (switching testing to testthat -- partial at this stage) and docs to
      roxygen format (partial at this stage).

   *  Changes to package structure to comply with new Rcpp standards

                CHANGES IN phylobase VERSION 0.6.5

   *  Updates from cout/cerr to Rcpp::Rcout/Rcerr

   *  Comments in Nexus tree strings are being removed before being processed by
      readNCL

                CHANGES IN phylobase VERSION 0.6.3

   *   Fixed bugs in getNode in cases where labels included regexpr
       metacharacters and when a tip was labelled 0

   *   New methods: depthTips, nodeDepth and isUltrametric


                CHANGES IN phylobase VERSION 0.6.2

   *   Improve handling of errors returned by NCL (NxsException)

   *   Fix bug in case state labels are missing from the NEXUS file

   *   Upgrade to NCL 2.1.14

                CHANGES IN phylobase VERSION 0.6.1

   *   Fix bugs that prevented building on Windows 64-bit systems


                CHANGES IN phylobase VERSION 0.6

MAJOR CHANGES

    *   Updated to the Nexus Class Library (NCL) 2.1.12.

    *   Changed the way NCL is built during the installation process.

    *   Complete rewrite of the function readNexus which brings many new
        functionalities.

    *   Nodes labels do not have to be unique.


NEW FEATURES

    *   In readNexus, the option return.labels gives the state labels of the
        characters.

    *   It is now possible to import several types of  data blocks in a single
        NEXUS file with readNexus.

    *   The function phylobase.options() provides global options to control the
        behavior of the phylo4/phylo4d validator.

    *   The new method hasDuplicatedLabels() indicates whether any node labels
        are duplicated.

    *   The new method nData() returns the number of datasets associated with
        a tree.

    *   The column that contains the labels can now be specified by its name in
        the function formatData()

CHANGES

    *   The function getNode() has been modified to allow node matching in the
        case of non-unique labels.

    *   Many new unit tests.

BUG FIXES

    *   Far too many to document.  See the SVN log for details.

KNOWN ISSUES

    *   Unrooted trees are not supported by all functions, e.g. plot() and
        reorder().

    *   Factors are not supported by the default plotting method.


        *************************************************
        *                                               *
        *               0.5 SERIES NEWS                 *
        *                                               *
        *************************************************

                CHANGES IN phylobase VERSION 0.5

MAJOR CHANGES

    *   A var-cov matrix tree class, phylo4vcov, and methods for converting to
        and from other classes now exists.

    *   Replaced separate the tip.label and node.label slots with a unified
        label slot in the phylo4 class definition.

    *   Replaced separate the tip.data and node.data into a single data slot in
        the phylo4d class definition.

    *   The phylo4 class grew a annotate slot.

    *   The phylo4d class grew a metadata slot.

    *   Added an order slot to the phylo4 class definition and updated as()
        methods to assign the proper order (if possible) when converting
        between ape and phylobase tree formats.

    *   The Nnode slot was removed from the phylo4 class definition.

    *   An explicit root edge has been added to the edge matrix with 0 as the
        ancestor and nTips(phy) + 1 as the rood node.

    *   The edgeLabels() and edgeLength() accessors now return vectors with
        named elements in the same order as rows of the edge matrix, even when
        some/all values are missing.

    *   The labels() accessor and nodeID() methods now always return labels in
        ascending order of node ID

    *   Many function and argument names and defaults have been changed to make
        them more consistent most functions follow the getNode() pattern.

    *   The plotting functions have been replaced (see below).

    *   Now, data are matched against node numbers instead of node labels.

    *   Tip and internal node labels have now internal names that are character
        strings of the node number they correspond to. Thus it is possible to
        store labels in any order and assignment of labels more robust.

    *   We now use the RUnit package (not required for normal use) for adding
        unit tests.  Adding unit tests to inst/unitTests/ is now preferred over
        the tests/ directory.

    *   Numerous changes to pruning and tree subsetting code.  It is
        considerably more robust and no longer relies on calls to APE.

NEW FEATURES


R/subset.R
################
## subset phylo4
################

##' Methods for creating subsets of phylogenies
##' 
##' Methods for creating subsets of phylogenies, based on pruning a
##' tree to include or exclude a set of terminal taxa, to include all
##' descendants of the MRCA of multiple taxa, or to return a subtree
##' rooted at a given node.
##' 
##' The \code{subset} methods must be called using no more than one of
##' the four main subsetting criteria arguments (\code{tips.include},
##' \code{tips.exclude}, \code{mrca}, or \code{node.subtree}).  Each
##' of these arguments can be either character or numeric.  In the
##' first case, they are treated as node labels; in the second case,
##' they are treated as node numbers.  For the first two arguments,
##' any supplied tips not found in the tree (\code{tipLabels(x)}) will
##' be ignored, with a warning.  Similarly, for the \code{mrca}
##' argument, any supplied tips or internal nodes not found in the
##' tree will be ignored, with a warning.  For the \code{node.subtree}
##' argument, failure to provide a single, valid internal node will
##' result in an error.
##' 
##' Although \code{prune} is mainly intended as the workhorse function
##' called by \code{subset}, it may also be called directly.  In
##' general it should be equivalent to the \code{tips.exclude} form of
##' \code{subset} (although perhaps with less up-front error
##' checking).
##' 
##' The "[" operator, when used as \code{x[i]}, is similar to the
##' \code{tips.include} form of \code{subset}.  However, the indices
##' used with this operator can also be logical, in which case the
##' corresponding tips are assumed to be ordered as in \code{nodeId(x,
##' "tip")}, and recycling rules will apply (just like with a vector
##' or a matrix).  With a \linkS4class{phylo4d} object 'x',
##' \code{x[i,j]} creates a subset of \code{x} taking \code{i} for a
##' tip index and \code{j} for the index of data variables in
##' \code{tdata(geospiza, "all")}.  Note that the second index is
##' optional: \code{x[i, TRUE]}, \code{x[i,]}, and \code{x[i]} are all
##' equivalent.
##' 
##' Regardless of which approach to subsetting is used, the argument
##' values must be such that at least two tips are retained.
##' 
##' If the most recent common ancestor of the retained tips is not the
##' original root node, then the root node of the subset tree will be
##' a descendant of the original root.  For rooted trees with non-NA
##' root edge length, this has implications for the new root edge
##' length.  In particular, the new length will be the summed edge
##' length from the new root node back to the original root (including
##' the original root edge).  As an alternative, see the examples for
##' a way to determine the length of the edge that was immediately
##' ancestral to the new root node in the original tree.
##' 
##' Note that the correspondance between nodes and labels (and data in
##' the case of \linkS4class{phylo4d}) will be retained after all
##' forms of subsetting.  Beware, however, that the node numbers (IDs)
##' will likely be altered to reflect the new tree topology, and
##' therefore cannot be compared directly between the original tree
##' and the subset tree.
##' 
##' @name subset-methods
##' @docType methods
##' @param x an object of class \code{"phylo4"} or \code{"phylo4d"}
##' @param tips.include A vector of tips to include in the subset tree
##' @param tips.exclude A vector of tips to exclude from the subset
##' tree
##' @param mrca A vector of nodes for determining the most recent
##' common ancestor, which is then used as the root of the subset tree
##' @param node.subtree A single internal node specifying the root of
##' the subset tree
##' @param trim.internal A logical specifying whether to remove
##' internal nodes that no longer have tip descendants in the subset
##' tree
##' @param i (\code{[} method) An index vector indicating tips to
##' include
##' @param j (\code{[} method, phylo4d only) An index vector
##' indicating columns of node/tip data to include
##' @param \dots additional arguments to be passed to other methods
##' @return an object of class \code{"phylo4"} or \code{"phylo4d"}
##' @section Methods: \describe{ \item{x = "phylo4"}{subset tree}
##' \item{x = "phylo4d"}{subset tree and corresponding node and tip
##' data} }
##' @author Jim Regetz \email{regetz@@nceas.ucsb.edu}\cr Steven Kembel
##' \email{skembel@@berkeley.edu}\cr Damien de Vienne
##' \email{damien.de-vienne@@u-psud.fr}\cr Thibaut Jombart
##' \email{jombart@@biomserv.univ-lyon1.fr}
##' @keywords methods
##' @export
##' @rdname subset-methods
##' @examples
##' data(geospiza)
##' nodeLabels(geospiza) <- paste("N", nodeId(geospiza, "internal"), sep="")
##' geotree <- extractTree(geospiza)
##' 
##' ## "subset" examples
##' tips <- c("difficilis", "fortis", "fuliginosa", "fusca", "olivacea",
##'     "pallida", "parvulus", "scandens")
##' plot(subset(geotree, tips.include=tips))
##' plot(subset(geotree, tips.include=tips, trim.internal=FALSE))
##' plot(subset(geotree, tips.exclude="scandens"))
##' plot(subset(geotree, mrca=c("scandens","fortis","pauper")))
##' plot(subset(geotree, node.subtree=18))
##' 
##' ## "prune" examples (equivalent to subset using tips.exclude)
##' plot(prune(geotree, tips))
##' 
##' ## "[" examples (equivalent to subset using tips.include)
##' plot(geotree[c(1:6,14)])
##' plot(geospiza[c(1:6,14)])
##' 
##' ## for phylo4d, subset both tips and data columns
##' geospiza[c(1:6,14), c("wingL", "beakD")]
##' 
##' ## note handling of root edge length:
##' edgeLength(geotree)['0-15'] <- 0.1
##' geotree2 <- geotree[1:2]
##' ## in subset tree, edge of new root extends back to the original root
##' edgeLength(geotree2)['0-3']
##' ## edge length immediately ancestral to this node in the original tree
##' edgeLength(geotree, MRCA(geotree, tipLabels(geotree2)))
setGeneric("subset")

##' @rdname subset-methods
##' @aliases subset,phylo4-method
setMethod("subset", "phylo4", function(x, tips.include=NULL,
    tips.exclude=NULL, mrca=NULL, node.subtree=NULL, ...) {
    ##  FIXME: could eliminate NULL and make the test
    ## if (!missing) rather than if (!is.null)
    ## (might have to change next line?)
    if (sum(!sapply(list(tips.include, tips.exclude, mrca,
        node.subtree), is.null))>1) {
        stop("must specify at most one criterion for subsetting")
    }
    all.tips <- nodeId(x, "tip")
    if (!is.null(tips.include)) {
        nodes <- getNode(x, tips.include, missing="OK")
        is.valid.tip <- nodes %in% all.tips
        kept <- nodes[is.valid.tip]
        dropped <- setdiff(all.tips, kept)
        unknown <- tips.include[!is.valid.tip]
    } else if (!is.null(tips.exclude)) {
        nodes <- getNode(x, tips.exclude, missing="OK")
        is.valid.tip <- nodes %in% all.tips
        dropped <- nodes[is.valid.tip]
        kept <- setdiff(all.tips, dropped)
        unknown <- tips.exclude[!is.valid.tip]
    } else if (!is.null(mrca)) {
        nodes <- getNode(x, mrca, missing="OK")
        is.valid.node <- nodes %in% nodeId(x, "all")
        mnode <- MRCA(x, nodes[is.valid.node])
        if (length(mnode)!=1) {
            stop("mrca must include at least one valid node")
        }
        kept <- descendants(x, mnode)
        dropped <- setdiff(all.tips, kept)
        unknown <- mrca[!is.valid.node]
    } else if (!is.null(node.subtree)) {
        node <- getNode(x, node.subtree, missing="OK")
        if (length(node)!=1 || !(node %in% nodeId(x, "internal"))) {
            stop("node.subtree must be a single valid internal node")
        }
        kept <- descendants(x, node)
        dropped <- setdiff(all.tips, kept)
        unknown <- numeric(0)
    } else {
        kept <- getNode(x, nodeId(x, "tip"))
        dropped <- numeric(0)
        unknown <- numeric(0)
    }
    if (length(unknown)>0) {
        warning("invalid nodes ignored: ", paste(unknown, 
            collapse=", "))
    }
    if (length(kept)<2) {
        stop("0 or 1 tips would remain after subsetting")
    }
    if (length(dropped)==0) return(x)
    return(prune(x, dropped, ...))
})

###############
# '[' operator
###############

## Consider using some combination of these for stricter argument
## checking? Not implementing now because extra arguments are just
## ignored, which is fairly common S4 method behavior:
## * in "[" methods for phylo4:
##    if (nargs()>2) stop("unused arguments")
## * in "[" methods for both phylo4 and phylo4d:
##    if (!missing(...)) stop("unused argument(s)")

##### --------  phylo4 '[' methods

##' @rdname subset-methods
##' @aliases [,phylo4,character,missing-method
setMethod("[", signature(x="phylo4", i="character", j="missing",
    drop="missing"), function(x, i, j, ..., drop) {
    subset(x, tips.include=i)
})

##' @rdname subset-methods
##' @aliases [,phylo4,numeric,missing-method
setMethod("[", signature(x="phylo4", i="numeric", j="missing",
    drop="missing"), function(x, i, j, ..., drop) {
    subset(x, tips.include=i)
})

##' @rdname subset-methods
##' @aliases [,phylo4,logical,missing-method
setMethod("[", signature(x="phylo4", i="logical", j="missing",
    drop="missing"), function(x, i, j, ..., drop) {
    subset(x, tips.include=nodeId(x, "tip")[i])
})

##' @rdname subset-methods
##' @aliases [,phylo4,missing,missing-method
setMethod("[", signature(x="phylo4", i="missing", j="missing",
    drop="missing"), function(x, i, j, ..., drop) {
    x
})

##### --------  phylo4d '[' methods

##' @rdname subset-methods
##' @aliases [,phylo4d,ANY,character,missing-method
setMethod("[", signature(x="phylo4d", i="ANY", j="character",
    drop="missing"), function(x, i, j, ..., drop) {
    if (!missing(i)) x <- x[i]
    tdata(x, type="all") <- tdata(x, type="all")[j]
    return(x)
})

##' @rdname subset-methods
##' @aliases [,phylo4d,ANY,numeric,missing-method
setMethod("[", signature(x="phylo4d", i="ANY", j="numeric",
    drop="missing"), function(x, i, j, ..., drop) {
    if (!missing(i)) x <- x[i]
    tdata(x, type="all") <- tdata(x, type="all")[j]
    return(x)
})

##' @rdname subset-methods
##' @aliases [,phylo4d,ANY,logical,missing-method
setMethod("[", signature(x="phylo4d", i="ANY", j="logical",
    drop="missing"), function(x, i, j, ..., drop) {
    if (!missing(i)) x <- x[i]
    tdata(x, type="all") <- tdata(x, type="all")[j]
    return(x)
})

## borrow from Matrix package approach of trapping invalid usage
##' @rdname subset-methods
##' @aliases [,phylo4,ANY,ANY,ANY-method
setMethod("[", signature(x="phylo4", i="ANY", j="ANY", drop="ANY"),
    function(x, i, j, ..., drop) {
    stop("invalid argument(s)")
})

##### -------- prune

##' @rdname subset-methods
##' @aliases prune
setGeneric("prune", function(x, ...) {
    standardGeneric("prune")
})

## return characters, sorted in NUMERIC order
.chnumsort <- function(x) {
  as.character(sort(as.numeric(x)))
}

##' @rdname subset-methods
##' @aliases prune,phylo4-method
setMethod("prune", "phylo4",
    function(x, tips.exclude, trim.internal=TRUE) {

    makeEdgeNames <- function(edge) {
        paste(edge[,1], edge[,2], sep="-")
    } 

    ## drop tips and obsolete internal nodes from edge matrix
    tip.drop <- getNode(x, tips.exclude, missing="fail")
    tip.keep <- setdiff(nodeId(x, "tip"), tip.drop)
    nodes <- nodeId(x, "all")
    node.keep <- rep(FALSE, length(nodes))
    node.keep[tip.keep] <- TRUE
    if (trim.internal) {
        if (edgeOrder(x) == "postorder") {
            edge.post <- edges(x)
        } else {
            edge.post <- edges(reorder(x, "postorder"))
        }
        for (i in seq_along(edge.post[,2])) {
            if (node.keep[edge.post[i,2]]) {
                node.keep[edge.post[i,1]] <- TRUE
            }
        }
    } else {
        node.keep[nodeId(x, "internal")] <- TRUE
    }
    edge.new <- edges(x)[edges(x)[,2] %in% nodes[node.keep], ]
  
    ## remove singletons
    edge.length.new <- edgeLength(x)
    edge.label.new <- edgeLabels(x)
    singletons <- which(tabulate(edge.new[edge.new[, 1] != 0, 1])==1)
    while (length(singletons)>0) {
        sing.node <- singletons[1]

        ## update edge matrix
        edges.drop <- which(edge.new==sing.node, arr.ind=TRUE)[,"row"]
        sing.edges <- edge.new[edges.drop,]
        edge.new[edges.drop[2], ] <- c(sing.edges[2,1], sing.edges[1,2])
        edge.new <- edge.new[-edges.drop[1], ]

        ## update edge lengths and edge labels
        edge.names.drop <- makeEdgeNames(sing.edges)
        edge.name.new <- paste(sing.edges[2,1], sing.edges[1,2], sep="-")
        edge.length.new[edge.name.new] <-
            sum(edge.length.new[edge.names.drop])
        edge.length.new <- edge.length.new[-match(edge.names.drop,
            names(edge.length.new))]
        edge.label.new[edge.name.new] <- NA
        edge.label.new <- edge.label.new[-match(edge.names.drop,
            names(edge.label.new))]

        singletons <- which(tabulate(edge.new[edge.new[, 1] != 0, 1])==1)
    }

    ## remove dropped elements from labels
    label.new <- labels(x)[names(labels(x)) %in% edge.new]

    ## subset and order edge.length and edge.label with respect to edge
    edge.names <- makeEdgeNames(edge.new)
    edge.length.new <- edge.length.new[edge.names]
    edge.label.new <- edge.label.new[edge.names]

    if (!trim.internal) {
        ## make sure now-terminal internal nodes are treated as tips
        tip.now <- setdiff(edge.new[,2], edge.new[,1])
        tip.add <- tip.now[tip.now>nTips(x)]
        if (length(tip.add)>0) {
            ind <- match(tip.add, names(label.new))

            ## node renumbering workaround to satisfy plot method
            newid <- sapply(tip.add, function(tip) descendants(x, tip)[1])
            names(label.new)[ind] <- newid
            edge.new[match(tip.add, edge.new)] <- newid
            tip.now[match(tip.add, tip.now)] <- newid

            isTip <- edge.new %in% tip.now
            edge.new[isTip] <- match(edge.new[isTip],
            sort(unique.default(edge.new[isTip])))
        }
    }

    ## renumber nodes in the edge matrix
    edge.new[] <- match(edge.new, sort(unique.default(edge.new))) - 1L

    ## update corresponding element names in the other slots
    edge.names <- makeEdgeNames(edge.new)
    names(edge.length.new) <- edge.names
    names(edge.label.new) <- edge.names
    label.new <- label.new[order(as.numeric(names(label.new)))]
    names(label.new) <- seq_along(label.new)

    ## update, check, then return the pruned phylo4 object
    x@edge <- edge.new
    ##TODO would prefer to leave out NA from edge.length slot, but can't 
    x@edge.length <- edge.length.new
    x@edge.label <- edge.label.new[!is.na(edge.label.new)]
    x@label <- label.new[!is.na(label.new)]
    if(is.character(checkval <- checkPhylo4(x))) {
        stop(checkval)
    } else {
        return(x)
    }

})

##' @rdname subset-methods
##' @aliases prune,phylo4d-method
setMethod("prune", "phylo4d",
    function(x, tips.exclude, trim.internal=TRUE) {

    tree <- extractTree(x)
    phytr <- prune(tree, tips.exclude, trim.internal)

    ## create temporary phylo4 object with complete and unique labels
    tmpLbl <- .genlab("n", nTips(x)+nNodes(x))
    tmpPhy <- tree
    labels(tmpPhy, "all") <- tmpLbl
    tmpPhytr <- prune(tmpPhy, getNode(x, tips.exclude), trim.internal)

    ## get node numbers to keep
    oldLbl <- labels(tmpPhy, "all")
    newLbl <- labels(tmpPhytr, "all")
    wasKept <- oldLbl %in% newLbl
    nodesToKeep <- as.numeric(names(oldLbl[wasKept]))

    ## subset original data, and update names
    allDt <- x@data[match(nodesToKeep, rownames(x@data)), , drop=FALSE]
    rownames(allDt) <- match(newLbl, oldLbl[wasKept])

    phytr <- phylo4d(phytr, all.data=allDt, match.data=TRUE)

    phytr
})

## setMethod("prune","ANY",
##           function(phy, tip, trim.internal = TRUE, subtree = FALSE,
##                    ,...) {
##             if (class(phy)=="phylo") {
##               ape::prune(phy, tip, trim.internal, subtree)
##               } else stop("no prune method available for",
##                     deparse(substitute(phy)),
##                     "(class",class(phy),")")
##           })

